iceoryx_hoofs/source/posix_wrapper/shared_memory_object/shared_memory.cpp

```diff
@@ -125,46 +128,65 @@ int32_t SharedMemory::getHandle() const noexcept
     return m_handle;
 }
 
-bool SharedMemory::open(const int oflags, const mode_t permissions, const uint64_t size) noexcept
+bool SharedMemory::hasOwnership() const noexcept
 {
-    cxx::Expects(static_cast<int64_t>(size) <= std::numeric_limits<int64_t>::max());
+    return m_hasOwnership;
+}
 
+bool SharedMemory::open(const AccessMode accessMode,
+                        const Policy policy,
+                        const mode_t permissions,
+                        const uint64_t size) noexcept
+{
+    cxx::Expects(size <= static_cast<uint64_t>(std::numeric_limits<int64_t>::max()));
+
+    m_hasOwnership =
+        (policy == Policy::EXCLUSIVE_CREATE || policy == Policy::PURGE_AND_CREATE || policy == Policy::OPEN_OR_CREATE);
 
     // the mask will be applied to the permissions, therefore we need to set it to 0
     mode_t umaskSaved = umask(0U);
     {
         cxx::GenericRAII umaskGuard([&] { umask(umaskSaved); });
 
-        // if we create the shm, cleanup old resources
-        if (oflags & O_CREAT)
+        if (policy == Policy::PURGE_AND_CREATE)
         {
-            posixCall(shm_unlink)(m_name.c_str())
-                .failureReturnValue(-1)
-                .ignoreErrnos(ENOENT)
-                .evaluate()
-                .and_then([this](auto& r) {
-                    if (r.errnum != ENOENT)
-                    {
-                        std::cout << "SharedMemory still there, doing an unlink of " << m_name << std::endl;
-                    }
-                });
+            IOX_DISCARD_RESULT(posixCall(shm_unlink)(m_name.c_str())
+                                   .failureReturnValue(INVALID_HANDLE)
+                                   .ignoreErrnos(ENOENT)
+                                   .evaluate());
         }
 
-        if (posixCall(iox_shm_open)(m_name.c_str(), oflags, permissions)
-                .failureReturnValue(-1)
-                .evaluate()
-                .and_then([this](auto& r) { m_handle = r.value; })
-                .or_else([this](auto& r) { m_errorValue = this->errnoToEnum(r.errnum); })
-                .has_error())
+        auto result = posixCall(iox_shm_open)(m_name.c_str(), getOflagsFor(accessMode, policy), permissions)
+                          .failureReturnValue(INVALID_HANDLE)
+                          .suppressErrorMessagesForErrnos((policy == Policy::OPEN_OR_CREATE) ? EEXIST : 0)
+                          .evaluate();
+        if (result.has_error())
         {
+            // if it was not possible to create the shm exclusively someone else has the
+            // ownership and we just try to open it
+            if (policy == Policy::OPEN_OR_CREATE && result.get_error().errnum == EEXIST)
+            {
+                result = posixCall(iox_shm_open)(m_name.c_str(), getOflagsFor(accessMode, Policy::OPEN), permissions)
+                             .failureReturnValue(INVALID_HANDLE)
+                             .evaluate();
```

Comment:
I wouldn't say that this is an error since the whole object is then in an invalid state. But you are right we should be extra cautious and I change this.